File: python/sort/quicksort.py
ARR = []
gc = 1

def PARTITION(l, r):
    global ARR
    pivot = ARR[r]
    i = l-1
    for j in range(l, r):
        if ARR[j] < pivot:
            i += 1
            ARR[i], ARR[j] = ARR[j], ARR[i]
    ARR[i+1], ARR[r] = ARR[r], ARR[i+1]
    return i+1

def QS(l, r):
    global gc
    if r-1 <= l:
        return
    p = PARTITION(l, r-1)
    #print "After call #"+str(gc)+" to PARTITION procedure: ", ARR
    #gc += 1
    QS(l, p)
    QS(p+1, r)

def QUICKSORT(A):
    global ARR
    global gc
    gc = 1
    ARR = A
    QS(0, len(A))
    return ARR

File: java/sort/QuickSort.java
class QuickSort
{
  public static void quickSort(int a[])
  {
    qsort(a,0,a.length-1);
  }
  private static void qsort(int a[], int l, int r)
  {
    if(l<r)
    {
      int i=partition(a,l,r);
      qsort(a,l,i-1);
      qsort(a,i+1,r);
    }
  }
  private static int partition(int a[], int l, int r)
  {
    int i=l-1,j,t;
    for(j=l;j<r;j++)
    {
      if(a[j]<a[r])
      {
        i++;
        t=a[i];
        a[i]=a[j];
        a[j]=t;
      }
    }
    t=a[i+1];
    a[i+1]=a[r];
    a[r]=t;
    return i+1;
  }
}
